### PR TITLE
stored: add dedupable backend

### PR DESCRIPTION
**Pull request by sebsura**
### Thank you for contributing to the Bareos Project!

The dedup backend makes it possible for deduplicating filesystems to deduplicate backed up data.

Based of pr #1662 

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)
- [Boy Scout Rule](https://docs.bareos.org/DeveloperGuide/generaldevel.html#boy-scout-rule)

### Checklist for the _reviewer_ of the PR (will be processed by the Bareos team)
Make sure you check/merge the PR using `devtools/pr-tool` to have some simple automated checks run and a proper changelog record added.

##### General
- [x] Is the PR title usable as CHANGELOG entry?
- [x] Purpose of the PR is understood
- [x] Commit descriptions are understandable and well formatted
~~Check backport line~~
- [x] Required backport PRs have been created

##### Source code quality
- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR

##### Tests
- [x] Decision taken that a test is required (if not, then remove this paragraph)
- [x] The choice of the type of test (unit test or systemtest) is reasonable
- [x] Testname matches exactly what is being tested
- [x] On a fail, output of the test leads quickly to the origin of the fault

### CODE REVIEW

**Comment by sebsura** [comment]
One thing i was not sure about was the error reporting.  Currently the device itself outputs error messages whenever they are caught and prints an Emsg with it.  This makes it easier to give good error reports but there will probably be a lot of double reports ala:
```
dedup device : could not open dedup volume. ERR=...
normal bareos device: could not open dedup volume. ERR=...
```

An alternative would be to set errno appropriately (printing a dmsg instead) and let the normal bareos routines handle the error reporting.

The problem is that some problems do not have an associated error number since they might be internal logic errors so there is no way to give a helpful message in that case.

**Review by pstorz — CHANGES_REQUESTED** [review]
Please see comments. I think we should rename the device to `dedupable`

File: core/src/stored/backends/dedup_device.cc [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+    {
+      if (vol) {
+        vol->AbortBlock(std::move(save));
+        vol = nullptr;
+      }
+    }
+  };
+
+  if (!openvol) {
+    Emsg0(M_ERROR, 0, T_("Trying to write dedup volume when none are open.\n"));
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/stored/backends/dedup_device.cc:187]: This error should signal that you were trying to write to the device, but no volume was opened yet.  Ill try to see if i can rewrite it to make it clearer.

File: core/src/stored/backends/util.cc [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+      return format_parse_error_at("key is empty", str, pair);
+    }
+
+    if (val.size() == 0) {
+      return format_parse_error_at("val is empty", str, pair);
+    }
+
+    vec.emplace_back(key, val);
+
+    // if (auto [miter, inserted] = parsed.emplace(key, val); !inserted) {
```
[the remark this answers is not in the document]
Comment (sebsura) [reply core/src/stored/backends/util.cc:214]: Good catch.  The duplicate checking moved further to the bottom of the function, so this code is not necessary anymore.

File: docs/manuals/source/DeveloperGuide/mediaformat.rst [hunk 3]
```diff
@@ -592,7 +580,7 @@ Overall Storage Format
        |-------------------------------------------------------|
        |              JobLevel         (uint32_t)              |
        |-------------------------------------------------------|
-       |              FileSetMD5       (50 bytes)   VerNum 11  |
+       |              FileSetMD5       (128 bytes)             |
```
Comment (pstorz) [inline docs/manuals/source/DeveloperGuide/mediaformat.rst:583]: Is the 50 Bytes wrong in the old docs?

File: docs/manuals/source/DeveloperGuide/mediaformat.rst [hunk 4]
```diff
@@ -610,19 +598,17 @@ Overall Storage Format
        |-------------------------------------------------------|
        |              JobErrors        (uint32_t)              |
        |-------------------------------------------------------|
-       |              JobStatus        (uint32_t) VerNum 11    |
+       |              JobStatus        (uint32_t)              |
        :=======================================================:
        * => fields deprecated
```
Comment (pstorz) [inline docs/manuals/source/DeveloperGuide/mediaformat.rst:603]:
```suggestion
```

**Review by pstorz — CHANGES_REQUESTED** [review]
Please check for docs changes. I also would remove the remaining sentence about the deprecated bacula format in mediaformat.rst:

`From Bacula >= 1.27 and therefore Bareos, each block contains only records of a single job.`

File: docs/manuals/source/DeveloperGuide/mediaformat.rst (hunk 3 above)
Comment (sebsura) [reply docs/manuals/source/DeveloperGuide/mediaformat.rst:582]: Yes and no.
```
struct Session_Label {
...
  char FileSetMD5[MAX_NAME_LENGTH]{};
};
```
So the md5 can have up to 127+1 bytes, but our `SerialString` routine only writes out the actual string data.
As Md5 is guaranteed to be 32 + 1 bytes long, i guess the real answer is 33 bytes.  Im not sure where the 50bytes comes from though.

File: docs/manuals/source/DeveloperGuide/mediaformat.rst (hunk 4 above)
Comment (sebsura) [reply docs/manuals/source/DeveloperGuide/mediaformat.rst:602]: Done

File: docs/manuals/source/TasksAndConcepts/StorageBackends.rst [hunk 5]
```diff
@@ -375,3 +377,21 @@ for Bareos integration. You can use following snippet to configure it as storage
...
+
+.. _SdBackendDedup:
+
+Dedup Backend
+-------------
+
+**Dedup**
+
+A dedup (friendly) device splits up the data it receives into bareos specific data and file data
+to store them seperately.  It tries to store the file data blocks it receives only at certain intervals
```
[the remark this answers is not in the document]
Comment (sebsura) [reply docs/manuals/source/TasksAndConcepts/StorageBackends.rst:389]: Done

File: debian/control [hunk 6]
```diff
@@ -236,6 +236,19 @@ Description: Backup Archiving Recovery Open Sourced - storage daemon FIFO backen
...
 
+Package:        bareos-storage-dedup
+Architecture:   any
+Pre-Depends:    debconf (>= 1.4.30) | debconf-2.0
+Depends:        bareos-common (= ${binary:Version}), bareos-storage (= ${binary:Version}), lsb-base (>= 3.2-13), ${shlibs:Depends}, ${misc:Depends}
+Description: Backup Archiving Recovery Open Sourced - storage daemon DEDUP backend
+ Bareos is a set of programs to manage backup, recovery and verification of
+ data across a network of computers of different kinds.
+ .
+ This package contains the Storage backend for DEDUP files.
```
[the remark this answers is not in the document]
Comment (sebsura) [reply debian/control:247]: Done

File: core/platforms/packaging/bareos.spec
[the remark this answers is not in the document]
Comment (sebsura) [reply core/platforms/packaging/bareos.spec:1]: Agree

**Review by pstorz — APPROVED** [review]

**Pull request merged**